src/twisted/conch/test/test_ckeygen.py

```diff
@@ -628,3 +643,58 @@ def test_changePassphraseSubtypeV1(self):
         self.assertTrue(
             privateKeyContent.startswith(b"-----BEGIN OPENSSH PRIVATE KEY-----\n")
         )
+
+    def test_useDefaultForKey(self):
+        """
+        L{options} will default to "~/.ssh/id_rsa" if the user doesn't
+        specify a key.
+        """
+        input_prompts = []
+
+        def mock_input(*args):
+            return input_prompts.append("")
+
+        options = {"filename": ""}
+
+        filename = _getKeyOrDefault(options, mock_input)
+        self.assertEqual(
+            options["filename"],
+            "",
+        )
+        # Resolved path is an RSA key inside .ssh dir.
+        self.assertTrue(filename.endswith(os.path.join(".ssh", "id_rsa")))
+        # The user is prompted once to enter the path, since no path was
+        # provided via CLI.
+        self.assertEqual(1, len(input_prompts))
+        self.assertEqual([""], input_prompts)
+
+    def test_displayPublicKeyHandleFileNotFound(self):
+        """
+        Ensure FileNotFoundError is handled, whether the user has supplied
+        a bad path, or has no key at the default path.
+        """
+        options = {"filename": "/foo/bar"}
+        exc = self.assertRaises(SystemExit, displayPublicKey, options)
+        self.assertIn("could not be opened, please specify a file.", exc.args[0])
+
+    def test_changePassPhraseHandleFileNotFound(self):
+        """
+        Ensure FileNotFoundError is handled for an invalid filename.
+        """
+        options = {"filename": "/foo/bar"}
+        with self.assertRaises(SystemExit) as e:
+            changePassPhrase(options)
+            self.assertIn(
```

Comment:
Thanks for the updates and followp.


Just a remider. 
I see these tests calling assertion *after* an exception is raised.
In this case, the code after is never called... so the assertion is never raised.

I think this is verry bad habit 

My suggestion is to try TDD.

Write the test first without updating the implementation.
See that the test fails.
Update your implementation.
See the test pass.

If the test pass from the first run, is harder to tell if the test really does its job.

Cheers

Reply:
Updated to remove the context managers.